indent/ecrystal.vim

```diff
@@ -0,0 +1,487 @@
+
+  let [lnum, col] = searchpos(s:ecr_close, 'bWz', stop)
+
+  if !lnum
+    call cursor(pos)
+    return [open, close]
+  endif
+
+  let [open, close] = s:MatchECRControl()
+
+  while !open.lnum
+    let [lnum, col] = searchpos(s:ecr_close, 'bWz', stop)
+
+    if !lnum
+      break
+    endif
+
+    let [open, close] = s:MatchECRControl()
+  endwhile
+
+  call cursor(pos)
+  return [open, close]
+endfunction
+
+" GetEcrystalIndent {{{1
+" =================
+
+function GetEcrystalIndent() abort
+  let prev_lnum = prevnonblank(v:lnum - 1)
+
+  if b:ecrystal_indent_multiline
+    let [open_tag_lnum, open_tag_col] = s:MatchECR()
+  else
+    let open_tag_lnum = 0
+  endif
+
+  if open_tag_lnum && open_tag_lnum < v:lnum
+    " If we are inside a multiline eCrystal tag...
+
+    let ind = indent(open_tag_lnum)
+
+    " If this line has a closing delimiter that isn't inside a string or
+    " comment, then we are done with this tag
+    if crystal#indent#Match(v:lnum, s:ecr_close)
+      return ind
+    endif
+
+    " All tag contents will have at least one indent
+    let ind += s:sw()
+
+    if open_tag_lnum == prev_lnum
+      " If this is the first line after the opening delimiter, then one
+      " indent is enough
+      return ind
+    elseif s:MatchAt(open_tag_lnum, open_tag_col, s:ecr_comment_open)
+      " eCrystal comments shouldn't be indented any further
+      return ind
+    else
+      " Else, fall back to Crystal indentation...
+      let crystal_ind = GetCrystalIndent()
+
+      " But only if it isn't less than the default indentation for this
+      " tag
+      return crystal_ind < ind ? ind : crystal_ind
+    endif
+  else
+    let [prev_ecr_open, prev_ecr_close] = s:FindPrevECRControl(v:lnum, prev_lnum)
+    let [curr_ecr_open, curr_ecr_close] = s:FindECRControl()
+
+    let prev_is_ecr = prev_ecr_open.lnum != 0
+    let curr_is_ecr = curr_ecr_open.lnum != 0
+
+    let shift = 0
+
+    if prev_is_ecr
+      if s:ECRIndent(prev_ecr_open, prev_ecr_close)
+        let shift = 1
+      endif
+    endif
+
+    if curr_is_ecr
+      if s:ECRDedent(curr_ecr_open, curr_ecr_close)
+        if !b:ecrystal_align_end
+          let shift = shift ? 0 : -1
+        else
+          " Find the nearest previous control tag that starts an indent
+          " and align this one with that one
+          let end_tags = 0
+
+          let [open, close] = s:FindPrevECRControl()
+
+          while open.lnum
+            if s:ECRIndent(open, close)
+              if end_tags
+                let end_tags -= 1
+              else
+                return indent(open.lnum)
+              endif
+            elseif s:ECRDedent(open, close)
+              let end_tags += 1
+            endif
+
+            let [open, close] = s:FindPrevECRControl(open.lnum)
+          endwhile
+        endif
+      endif
+    endif
+
+    if shift
+      return indent(prev_lnum) + s:sw() * shift
+    else
+      exec 'return ' . b:ecrystal_subtype_indentexpr
```

Comment:
Please check the variable is not empty. `indentexpr` is user configuration and may be empty.

Reply:
This check is already made on line 15 (copied from vim-ruby).

Reply:
oh, ok. I did not consider that.